### PR TITLE
Add in info about custom iterators.

### PR DESCRIPTION
**Pull request by powerc9000**

### CODE REVIEW

File: content/docs/overview.md [hunk 1]
```diff
@@ -1792,6 +1792,51 @@ f: Foo(T, N);
...
+    value: int,
+    next: ^Node
+}
+
+MyIterator :: struct {
+  index: int,
+  current: Node
+}
+
+iterate :: proc(it: ^MyIterator) -> (value: Node, index: int, ok: bool) {
```
Comment (Tetralux) [inline content/docs/overview.md:1815]: Is it worth mentioning that the name is not important?

File: content/docs/overview.md [hunk 2]
```diff
@@ -1792,6 +1792,51 @@ f: Foo(T, N);
...
+
+main :: proc () {
+    head : Node = {value=10, next=nil};
+    tail : Node = {value=10, next=nil};
+    
+    head.next = tail;
+    
+    it : MyIterator = {index = 0; current=head};
+    
+    for node in iterate(&it) {
```
Comment (Tetralux) [inline content/docs/overview.md:1834]: Is it worth having a comment here to make it very clear that _iterating over the function_ is how you use it for this?

File: content/docs/overview.md [hunk 3]
```diff
@@ -1792,6 +1792,61 @@ f: Foo(T, N);
 #assert(size_of(f) == (N+N-2)*size_of(T));
 ```
 
+## Custom iterator functions
+
+Custom iterator functions can be used to iterate over anything you want.
```
Comment (gingerBill) [inline content/docs/overview.md:1797]: Use the word `procedures` not `functions`

File: content/docs/overview.md [hunk 4]
```diff
@@ -1792,6 +1792,61 @@ f: Foo(T, N);
...
+
+A custom iterator function should take a pointer to a struct keeps track of the state of the iteration.
```
Comment (gingerBill) [inline content/docs/overview.md:1799]: I'd rephrase this to:

```
It is customary for a custom iterator procedure to have a pointer to a value in order to keep track of the state of the iteration.
```

File: content/docs/overview.md [hunk 5]
```diff
@@ -1792,6 +1792,61 @@ f: Foo(T, N);
...
+
+The custom iterator function should return 3 values. The next value to be given to the loop, the next index for the loop, and a boolean to signify if the loop should keep running.
```
Comment (gingerBill) [inline content/docs/overview.md:1801]: Custom iterators can return 1, 2, or 3 values, as long as the last value is a boolean.

File: content/docs/overview.md [hunk 6]
```diff
@@ -1792,6 +1792,61 @@ f: Foo(T, N);
...
+    
+    head.next = tail;
+    
+    it : MyIterator = {index = 0; current=head};
+    
+    for node in custom_iterator_function(&it) {
+      fmt.println(node);
+    }
+}
+```
```
Comment (gingerBill) [inline content/docs/overview.md:1838]: I am not sure a linked list is the best example especially since honestly prefer linked lists to be C-style loops e.g.
```odin
for node := head; node != nil; node = node.next {
    ...
}
```

However, I am not sure what the best example would be.

File: content/docs/overview.md [hunk 7]
```diff
@@ -1792,6 +1792,61 @@ f: Foo(T, N);
...
+
+Using a custom iterator function is equivalent to:
```
Comment (gingerBill) [inline content/docs/overview.md:1840]: `Using the above custom iterator is equivalent to:`

File: content/docs/overview.md (hunk 4 above)
Comment (Tetralux) [reply content/docs/overview.md:1799]: Is "to a value" clear enough there? Like - `value` doesn't really convey anything about what it's about.

Comment (gingerBill) [reply content/docs/overview.md:1799]: Sure.

File: content/docs/overview.md (hunk 6 above)
Comment (Tetralux) [reply content/docs/overview.md:1838]: An iterator that visits the elements in a random order? Albeit not perhaps the _most_ realistic example...

**Comment by powerc9000** [comment]
Updated most of the points brought up.
